src/awkward/operations/ak_to_parquet_dataset.py

```diff
@@ -0,0 +1,115 @@
+            search for files recursively. Ignored if `filenames` is None.
+
+    Creates a `_common_metadata` and a `_metadata` in a directory of Parquet files.
+
+        >>> ak.to_parquet(array1, "/directory/arr1.parquet", parquet_compliant_nested=True)
+        >>> ak.to_parquet(array2, "/directory/arr2.parquet", parquet_compliant_nested=True)
+        >>> ak.to_parquet_dataset("/directory")
+
+    The `_common_metadata` contains the schema that all files share. (If the files
+    have different schemas, this function raises an exception.)
+
+    The `_metadata` contains row-group metadata used to seek to specific row-groups
+    within the multi-file dataset.
+    """
+
+    # Implementation
+
+    from os import fsdecode, path
+
+    import awkward._connect.pyarrow
+
+    pyarrow_parquet = awkward._connect.pyarrow.import_pyarrow_parquet(
+        "ak.to_parquet_dataset"
+    )
+
+    try:
+        directory = fsdecode(directory)
+    except TypeError:
+        raise TypeError(
+            f"'directory' argument of 'ak.to_parquet_dataset' must be a path-like, not {type(directory).__name__}"
+        ) from None
+
+    directory = _regularize_path(directory)
+    if not path.isdir(directory):
+        raise ValueError(
+            f"{directory!r} is not a local filesystem directory" + {__file__}
+        )
+
+    if filenames is None:
+        import glob
+
+        filenames = sorted(
+            glob.glob(directory + f"/**/*{filename_extension}", recursive=True)
+        )
+    else:
+        filenames = [_regularize_path(x) for x in filenames]
+        filenames = [x if path.isabs(x) else path.join(directory, x) for x in filenames]
+
+    relpaths = [path.relpath(x, directory) for x in filenames]
+    schema, metadata_collector = _common_parquet_schema(
+        pyarrow_parquet, filenames, relpaths
+    )
+    pyarrow_parquet.write_metadata(schema, path.join(directory, "_common_metadata"))
+    pyarrow_parquet.write_metadata(
+        schema,
+        path.join(directory, "_metadata"),
+        metadata_collector=metadata_collector,
+    )
+
+
+def _regularize_path(path):
+    import os
+
+    if isinstance(path, getattr(os, "PathLike", ())):
+        path = os.fspath(path)
+
+    elif hasattr(path, "__fspath__"):
+        path = os.fspath(path)
+
+    elif path.__class__.__module__ == "pathlib":
+        import pathlib
+
+        if isinstance(path, pathlib.Path):
+            path = str(path)
+
+    if isinstance(path, str):
+        path = os.path.expanduser(path)
+
+    return path
+
+
+def _common_parquet_schema(pq, filenames, relpaths):
+    assert len(filenames) != 0
+
+    schema = None
+    metadata_collector = []
+    for filename, relpath in zip(filenames, relpaths):
+        if schema is None:
+            schema = pq.ParquetFile(filename).schema_arrow
+            first_filename = filename
+        elif not schema.equals(pq.ParquetFile(filename).schema_arrow):
```

Comment:
Whilst not cleanly defined, parquet does in principle allow schema evolution, but perhaps not in conjunction with (_common)_metadata